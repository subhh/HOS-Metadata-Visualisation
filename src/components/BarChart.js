import React, { useMemo, useState } from 'react'
import { max, values, entries, orderBy } from 'lodash'
import { Flex } from 'rebass'
import { scaleLinear } from 'd3'
import styled from 'styled-components'
import {
  ChartContainer,
  Title,
  Table,
  Row,
  Bar,
  Label,
  Value,
  BarWrapper,
} from 'components/ChartAtoms'
import Link from 'components/Link'
import { format } from 'utils'

const Collapsible = styled.div`
  max-height: ${p => (p.collapse ? p.height + 'px' : 'unset')};
  overflow: hidden;
`

/**
 * Bar chart component
 */
const BarChart = ({ title = '', data }) => {
  const [expanded, setExpanded] = useState(false)

  return useMemo(() => {
    const entr = entries(data)
    const orderedEntries = orderBy(entr, 1, 'desc')
    const isLong = entr.length > 10
    const xScale = scaleLinear()
      .domain([0, max(values(data))])
      .range([2, 100])

    return (
      <ChartContainer>
        <Title>{title}</Title>
        <Collapsible collapse={isLong && !expanded} height={238}>
          <Table>
            <tbody>
              {orderedEntries.map(([key, value], i) => {
                return (
                  <Row key={`row-${i}`} h={4}>
                    <Label title={key} color="greyDark">
                      {key}
                    </Label>
                    <Value color="grey">{format(value)}</Value>
                    <BarWrapper>
                      <Bar width={`${xScale(value)}%`} height={2} />
                    </BarWrapper>
                  </Row>
                )
              })}
            </tbody>
          </Table>
        </Collapsible>
        {isLong && (
          <Flex justifyContent="center" mt={5}>
            <Link
              fontSize={0}
              fontWeight={500}
              onClick={() => setExpanded(!expanded)}
            >
              {expanded ? 'Weniger anzeigen' : 'Alles anzeigen'}
            </Link>
          </Flex>
        )}
      </ChartContainer>
    )
  }, [title, data, expanded])
}

export default BarChart
