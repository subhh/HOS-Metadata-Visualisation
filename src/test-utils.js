import React, { useReducer } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { MemoryRouter } from 'react-router-dom'
import { StateContext, DispatchContext, reducer, initialState } from './state'
import theme from './theme'

const AllTheProviders = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <ThemeProvider theme={theme}>
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={dispatch}>
          {children}
        </DispatchContext.Provider>
      </StateContext.Provider>
    </ThemeProvider>
  )
}

const WithRouter = ({ children }) => (
  <MemoryRouter>
    <AllTheProviders>{children}</AllTheProviders>
  </MemoryRouter>
)

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderWithRouter = (ui, options) =>
  render(ui, {
    wrapper: WithRouter,
    ...options,
  })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export const mockAsync = data => {
  jest.spyOn(global, 'fetch').mockImplementation(url => {
    return Promise.resolve({
      json: () => Promise.resolve(data),
    })
  })
}

export const restoreAsyncMock = () => {
  global.fetch.mockRestore()
}
