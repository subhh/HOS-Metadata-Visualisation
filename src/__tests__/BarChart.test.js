import React from 'react'
import { render } from 'test-utils'
import BarChart from 'components/BarChart'

it('<BarChart /> without data should match snapshot', () => {
  const { container } = render(<BarChart title="Bar chart title" />)
  expect(container).toMatchSnapshot()
})

it('<BarChart /> with data should match snapshot', () => {
  const { container } = render(
    <BarChart title="Bar chart title" data={{ var1: 10, var2: 30 }} />
  )
  expect(container).toMatchSnapshot()
})
