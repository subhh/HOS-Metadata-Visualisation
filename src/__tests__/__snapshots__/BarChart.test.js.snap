// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarChart /> with data should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-htoDjs bZvMhP"
  >
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb sc-dnqmqq fCKnkB"
      color="blueDark"
      font-size="2"
      font-weight="500"
    >
      Bar chart title
    </h2>
    <div
      class="sc-jKJlTe hcfcuU"
      height="238"
    >
      <table
        class="sc-iwsKbI XZnfg"
      >
        <tbody>
          <tr
            class="sc-gZMcBi kbFdoy"
          >
            <td
              class="sc-VigVT frYtZr"
              color="greyDark"
              title="var2"
            >
              var2
            </td>
            <td
              class="sc-jTzLTM bwfcDe"
              color="grey"
            >
              30
            </td>
            <td
              class="sc-fjdhpX hAUhPj"
            >
              <div
                class="sc-bdVaJa sc-gqjmRU dObLIX"
                height="2"
                width="100%"
              />
            </td>
          </tr>
          <tr
            class="sc-gZMcBi kbFdoy"
          >
            <td
              class="sc-VigVT frYtZr"
              color="greyDark"
              title="var1"
            >
              var1
            </td>
            <td
              class="sc-jTzLTM bwfcDe"
              color="grey"
            >
              10
            </td>
            <td
              class="sc-fjdhpX hAUhPj"
            >
              <div
                class="sc-bdVaJa sc-gqjmRU cIGiEG"
                height="2"
                width="34.666666666666664%"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`<BarChart /> without data should match snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa sc-htoDjs bZvMhP"
  >
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb sc-dnqmqq fCKnkB"
      color="blueDark"
      font-size="2"
      font-weight="500"
    >
      Bar chart title
    </h2>
    <div
      class="sc-jKJlTe hcfcuU"
      height="238"
    >
      <table
        class="sc-iwsKbI XZnfg"
      >
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
